pkg/cloud/tags/tags.go

```diff
@@ -54,9 +54,21 @@ func Apply(params *ApplyParams) error {
 
 // Ensure applies the tags if the current tags differ from the params.
 func Ensure(current infrav1.Tags, params *ApplyParams) error {
-	want := infrav1.Build(params.BuildParams)
-	if !current.Equals(want) {
+	diff := computeDiff(current, params.BuildParams)
+	if len(diff) > 0 {
 		return Apply(params)
 	}
 	return nil
 }
+
+func computeDiff(current infrav1.Tags, buildParams infrav1.BuildParams) infrav1.Tags {
+	want := infrav1.Build(buildParams)
+
+	// Some tags could be external set by some external entities
+	// and that means even if there is no change in cluster
+	// managed tags, tags would be updated as "current" and
+	// "want" would be different due to external tags.
+	// This fix makes sure that tags are updated only if
+	// there is a change in cluster managed tags.
+	return want.Difference(current)
+}
```

Comment:
Can we just move these 2 lines in the `Ensure` call instead?

Reply:
I'm guessing the reason he did it this way is to be able to unit test computeDiff. If you want to unit test the other func, you have to either mock out the ec2 CreateTags call, or we need to use something like a function pointer with test/default configurations to avoid ec2.

Reply:
Yeah makes sense, it's probably fine to keep it this way, although it's definitely harder to read/follow than before